include/groov/mmio_bus.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <async/concepts.hpp>
+#include <async/just.hpp>
+#include <async/then.hpp>
+
+#include <bit>
+#include <cstdint>
+
+namespace groov {
+namespace detail {
+template <std::size_t NumBits> struct uint_of_width;
+
+template <> struct uint_of_width<8> {
+    using type = std::uint8_t;
+};
+
+template <> struct uint_of_width<16> {
+    using type = std::uint16_t;
+};
+
+template <> struct uint_of_width<32> {
+    using type = std::uint32_t;
+};
+
+template <> struct uint_of_width<64> {
+    using type = std::uint64_t;
+};
+
+template <std::size_t NumBits>
+using uint_of_width_t = typename uint_of_width<NumBits>::type;
+} // namespace detail
+
+struct cpp_mem_iface {
+    template <std::unsigned_integral T> static auto store(std::uintptr_t addr) {
+        return async::then([=](T value) -> void {
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            *reinterpret_cast<T volatile *>(addr) = value;
+        });
+    }
+
+    template <std::unsigned_integral T>
+    static auto load(std::uintptr_t addr) -> async::sender auto {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        T value = *reinterpret_cast<T volatile *>(addr);
+        return async::just(value);
+    }
+};
+
+template <typename BaseType = std::uint32_t, typename AddrType = std::uintptr_t,
+          typename HardwareInterface = cpp_mem_iface>
+struct mmio_bus {
+    using base_type = BaseType;
+    using addr_type = AddrType;
+    using iface = HardwareInterface;
+
+    template <base_type Mask>
+    static auto write(addr_type addr, base_type value) -> async::sender auto {
+        constexpr std::size_t lsb = std::countr_zero(Mask);
+        constexpr std::size_t width = std::countr_one(Mask >> lsb);
+        constexpr bool byte_aligned = (lsb % 8) == 0;
+        constexpr bool contiguous = std::popcount(Mask) == width;
+        constexpr bool pow2_width = std::popcount(width) == 1;
+        constexpr bool accessible_width = width & (8u + 16u + 32u + 64u);
```

Comment:
suggestion (non-blocking): For me, using `or` makes the code more readable with the use of the bit-wise `and`.

```c++
constexpr bool accessible_width = width & (8u | 16u | 32u | 64u);
```

Reply:
agreed, made the change